"""Given a FASTA file generated by a multiple alignment tool, aggressively call for variants."""
import sys

import pysam

fasta_fh = pysam.FastaFile(sys.argv[1])
vcf_h = [
    "##fileformat=VCFv4.2",
]
vcf = []

seqs = {}
lengths = []
for reference in fasta_fh.references:
    curr_ref = fasta_fh.fetch(reference=reference)
    seqs[reference] = curr_ref
    lengths.append(len(curr_ref))

LEN = max(lengths)
#print len(seqs), seqs.keys(), LEN

SNPS = 0
for i in range(LEN):
    tides = {}
    for seq in seqs:
        tide = seqs[seq][i]
        if tide == "-":
            continue

        if tide not in tides:
            tides[tide] = 1
        else:
            tides[tide] += 1

    if len(tides) > 1:
        #print tides
        SNPS += 1
        vcf.append([
            "contig",
            i+1,
            '.',
            'G',
            'A,T,C',
            0,
            '.',
            "INFO"
        ])


for r in vcf_h:
    print r
for r in vcf:
    print "\t".join([str(s) for s in r])
